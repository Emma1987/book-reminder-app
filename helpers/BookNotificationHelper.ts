import { scheduleNotification } from '@/helpers/NotificationHelper';
import { NotificationType, BookType } from '@/types/types';

const createNotification = (book: BookType, trigger: Date): NotificationType => {
    return {
        id: `alert-book-${book.id}`,
        bookId: book.id,
        notificationIdentifier: null,
        scheduledTime: trigger,
        title: 'Book Release Alert ðŸš¨',
        content: `The book ${book.title} is released today!`,
        readAt: null,
    };
};

export const scheduleBookReleaseNotification = async (
    book: BookType,
    addNotification: (notification: NotificationType) => void,
): Promise<void> => {
    if (!book.releaseDateRaw) return;

    try {
        const releaseDate = new Date(book.releaseDateRaw);
        releaseDate.setHours(10, 30, 0);

        const notification = createNotification(book, releaseDate);
        const identifier = await scheduleNotification(
            notification.title,
            notification.content,
            notification.scheduledTime,
        );
        notification.notificationIdentifier = identifier;

        addNotification(notification);
    } catch (error) {
        console.error('Error scheduling notification:', error);
    }
};
