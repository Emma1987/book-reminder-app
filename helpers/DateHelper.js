import { parse, isBefore, startOfDay } from 'date-fns';

export const formatDateStr = (dateString: string): string | null => {
    const months = ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'];
    const fullDatePattern = /^\d{4}-\d{2}-\d{2}$/;
    const monthYearPattern = /^\d{4}-\d{2}$/;
    const yearPattern = /^\d{4}$/;
    
    if (fullDatePattern.test(dateString)) {
        const [year, month, day] = dateString.split('-');
        const monthName = months[parseInt(month, 10) - 1];

        return `${day} ${monthName} ${year}`;
    }

    if (monthYearPattern.test(dateString)) {
        const [year, month] = dateString.split('-');
        const monthName = months[parseInt(month, 10) - 1];

        return `${monthName} ${year}`;
    }

    if (yearPattern.test(dateString)) {
        return dateString;
    }

    return null;
};

export const getDateObject = (dateString: string): Date | null => {
    const fullDatePattern = /^\d{4}-\d{2}-\d{2}$/;
    const monthYearPattern = /^\d{4}-\d{2}$/;
    const yearPattern = /^\d{4}$/;

    if (fullDatePattern.test(dateString)) {
        return parse(dateString, 'yyyy-MM-dd', new Date());
    }

    if (monthYearPattern.test(dateString)) {
        return parse(dateString, 'yyyy-MM', new Date());
    }

    if (yearPattern.test(dateString)) {
        return parse(dateString, 'yyyy', new Date());
    }

    return null;
}
