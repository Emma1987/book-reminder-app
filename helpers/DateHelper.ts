import { parse } from 'date-fns';

export const formatDateStr = (
    dateString: string | null,
    formatType: 'dayFirst' | 'monthFirst' = 'dayFirst',
): string | null => {
    if (!dateString) return null;

    const fullDatePattern = /^\d{4}-\d{2}-\d{2}$/;
    const monthYearPattern = /^\d{4}-\d{2}$/;
    const yearPattern = /^\d{4}$/;

    try {
        if (fullDatePattern.test(dateString)) {
            const date = new Date(dateString);
            const day = date.getDate();
            const month = date.toLocaleString('en-US', { month: 'short' });
            const year = date.getFullYear();

            if (formatType === 'dayFirst') {
                return `${day} ${month} ${year}`;
            } else {
                return `${month} ${day}, ${year}`;
            }
        }

        if (monthYearPattern.test(dateString)) {
            const [year, month] = dateString.split('-');
            const date = new Date(Number(year), Number(month) - 1);

            return new Intl.DateTimeFormat('en-US', {
                month: 'short',
                year: 'numeric',
            }).format(date);
        }

        if (yearPattern.test(dateString)) {
            return dateString;
        }
    } catch (error) {
        console.error('Error formatting date:', error);
    }

    return null;
};

export const getDateObject = (dateString: string): Date | null => {
    const fullDatePattern = /^\d{4}-\d{2}-\d{2}$/;
    const monthYearPattern = /^\d{4}-\d{2}$/;
    const yearPattern = /^\d{4}$/;

    if (fullDatePattern.test(dateString)) {
        return parse(dateString, 'yyyy-MM-dd', new Date());
    }

    if (monthYearPattern.test(dateString)) {
        return parse(dateString, 'yyyy-MM', new Date());
    }

    if (yearPattern.test(dateString)) {
        return parse(dateString, 'yyyy', new Date());
    }

    return null;
};
