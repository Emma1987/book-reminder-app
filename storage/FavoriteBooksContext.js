import React, { createContext, useState, useEffect } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';

export const FavoriteBooksContext = createContext();

export const FavoriteBooksProvider = ({ children }) => {
    const [favorites, setFavorites] = useState([]);

    useEffect(() => {
        loadFavorites();
    }, []);

    const loadFavorites = async () => {
        const storedFavorites = await AsyncStorage.getItem('favorites');
        if (storedFavorites) setFavorites(JSON.parse(storedFavorites));
    };

    const addFavorite = async (book) => {
        const updatedFavorites = [...favorites, book];
        setFavorites(updatedFavorites);
        await AsyncStorage.setItem('favorites', JSON.stringify(updatedFavorites));
    };

    const removeFavorite = async (bookId) => {
        const updatedFavorites = favorites.filter((book) => book.id !== bookId);
        setFavorites(updatedFavorites);
        await AsyncStorage.setItem('favorites', JSON.stringify(updatedFavorites));
    };

    const isFavorite = (bookId) => favorites.some((book) => book.id === bookId);

    return (
        <FavoriteBooksContext.Provider value={{ favorites, addFavorite, removeFavorite, isFavorite }}>
            {children}
        </FavoriteBooksContext.Provider>
    );
};
