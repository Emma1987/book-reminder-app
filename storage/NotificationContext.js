import React, { createContext, setTimeout, useEffect, useState } from 'react';
import AsyncStorage from '@react-native-async-storage/async-storage';
import * as Notifications from 'expo-notifications';

export const NotificationContext = createContext();

export const NotificationProvider = ({ children }) => {
    const [unreadNotifications, setUnreadNotifications] = useState(0);
    const [notifications, setNotifications] = useState([]);

    const sortAndSetNotifications = (notifications) => {
        const sortedNotifications = notifications.sort((a, b) => new Date(b.scheduledTime) - new Date(a.scheduledTime));
        setNotifications(sortedNotifications);
    };

    const countUnreadNotifications = (notifications) => {
        const now = new Date();
        return notifications.filter((notification) => {
            return new Date(notification.scheduledTime) <= now && !notification.readAt;
        }).length;
    };

    useEffect(() => {
        loadNotifications();
    }, []);

    useEffect(() => {
        const unreadCount = countUnreadNotifications(notifications);
        setUnreadNotifications(unreadCount);
    }, [notifications]);

    useEffect(() => {
        const subscription = Notifications.addNotificationReceivedListener(async (notification) => {
            setTimeout(async () => {
                await loadNotifications();
            }, 5000);
        });

        return () => subscription.remove();
    }, []);

    const loadNotifications = async () => {
        try {
            const storedNotifications = await AsyncStorage.getItem('notifications');

            if (storedNotifications) {
                const parsedNotifications = JSON.parse(storedNotifications);
                sortAndSetNotifications([...parsedNotifications]);
            }
        } catch (error) {
            console.error('Error loading notifications:', error);
        }
    };

    const addNotification = async (notification) => {
        const updatedNotifications = [...notifications, notification];
        sortAndSetNotifications(updatedNotifications);
        await AsyncStorage.setItem('notifications', JSON.stringify(updatedNotifications));
    };

    const updateNotification = async (notificationId, updatedFields) => {
        const updatedNotifications = notifications.map((notification) =>
            notification.id === notificationId ? { ...notification, ...updatedFields } : notification,
        );

        sortAndSetNotifications(updatedNotifications);
        await AsyncStorage.setItem('notifications', JSON.stringify(updatedNotifications));
    };

    const removeNotification = async (notificationId) => {
        const updatedNotifications = notifications.filter((notification) => notification.id !== notificationId);
        sortAndSetNotifications(updatedNotifications);
        await AsyncStorage.setItem('notifications', JSON.stringify(updatedNotifications));
    };

    const getNotificationByBookId = (bookId) => notifications.find((notification) => notification.bookId === bookId);

    return (
        <NotificationContext.Provider
            value={{
                notifications,
                unreadNotifications,
                loadNotifications,
                addNotification,
                updateNotification,
                removeNotification,
                getNotificationByBookId,
            }}
        >
            {children}
        </NotificationContext.Provider>
    );
};
