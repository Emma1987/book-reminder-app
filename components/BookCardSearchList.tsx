import React, { useContext } from 'react';
import { Alert, Image, Keyboard, StyleSheet, Text, TouchableOpacity, View } from 'react-native';
import { Ionicons } from '@expo/vector-icons';

import { ThemedText } from '@/components/ThemedText';
import { BookType } from '@/components/types';
import { Colors } from '@/constants/Colors';
import { isPublished } from '@/helpers/BookHelper';
import { formatDateStr } from '@/helpers/DateHelper';
import { scheduleBookReleaseNotification } from '@/helpers/BookNotificationHelper';
import { FavoriteBooksContext } from '@/storage/FavoriteBooksContext';
import { NotificationContext } from '@/storage/NotificationContext';

type BookCardSearchListProps = {
    book: BookType;
    onSeeDetails: (book: BookType) => void;
};

export function BookCardSearchList({ book, onSeeDetails }: BookCardSearchListProps) {
    const { addFavorite, removeFavorite, isFavorite } = useContext(FavoriteBooksContext);
    const { addNotification } = useContext(NotificationContext);

    const getNotified = () => {
        const alertContent = `ðŸ“š The book "${book.title}" is set to be released on ${formatDateStr(book.releaseDateRaw, 'monthFirst')}.\n\nWould you like to receive a notification on the release date?`;

        Alert.alert('Stay Updated!', alertContent, [
            {
                text: 'No, thanks',
                style: 'cancel',
                onPress: () => console.log('Notification declined'),
            },
            {
                text: 'Yes, notify me!',
                style: 'default',
                onPress: async () => {
                    scheduleBookReleaseNotification(book, addNotification);
                },
            },
        ]);
    };

    const toggleFavorite = () => {
        Keyboard.dismiss();

        if (isFavorite(book.id)) {
            removeFavorite(book.id);
        } else {
            if (!isPublished(book) && book.releaseDateRaw !== null) {
                getNotified();
            }

            addFavorite(book);
        }
    };

    return (
        <View style={styles.container}>
            {/* Book Cover */}
            <Image
                source={
                    book.coverImage ? { uri: book.coverImage } : require('@/assets/images/book-cover-placeholder.jpg')
                }
                style={styles.bookCover}
                accessibilityRole="image"
                accessibilityLabel="Book cover"
            />

            {/* Book Description */}
            <View style={styles.bookDescription}>
                <ThemedText type="subtitle">{book.title}</ThemedText>
                <ThemedText type="defaultSemiBold">{book.authors?.[0] ?? ''}</ThemedText>
                <Text>Release date: {formatDateStr(book.releaseDateRaw)}</Text>
            </View>

            {/* Action Icons */}
            <View style={styles.actions}>
                <TouchableOpacity
                    onPress={toggleFavorite}
                    accessibilityLabel="Add to favorites"
                    accessibilityHint="Adds this book to your favorites list"
                >
                    {isFavorite(book.id) && <Ionicons name="heart" size={25} color={Colors.red} />}
                    {!isFavorite(book.id) && <Ionicons name="heart-outline" size={25} color={Colors.gray} />}
                </TouchableOpacity>
                <TouchableOpacity
                    onPress={() => onSeeDetails(book)}
                    accessibilityLabel="View book details"
                    accessibilityHint={`View more details about ${book.title}`}
                >
                    <Ionicons name="information-circle-outline" size={25} color={Colors.gray} />
                </TouchableOpacity>
            </View>
        </View>
    );
}

const styles = StyleSheet.create({
    container: {
        flexDirection: 'row',
        alignItems: 'center',
        padding: 10,
        borderBottomWidth: 1,
        borderColor: Colors.white,
    },
    bookCover: {
        width: 60,
        height: 90,
    },
    bookDescription: {
        flex: 1,
        padding: 10,
    },
    actions: {
        width: 60,
        justifyContent: 'space-between',
        alignItems: 'center',
    },
});
