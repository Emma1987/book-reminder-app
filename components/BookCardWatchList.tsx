import React, { useContext } from 'react';
import { Alert, Image, StyleSheet, TouchableOpacity, View } from 'react-native';
import { Ionicons } from '@expo/vector-icons';

import { ThemedText } from '@/components/ThemedText';
import { BookType, NotificationType } from '@/types/types';
import { Colors } from '@/constants/Colors';
import { isPublished } from '@/helpers/BookHelper';
import { scheduleBookReleaseNotification } from '@/helpers/BookNotificationHelper';
import { formatDateStr } from '@/helpers/DateHelper';
import i18n from '@/i18n/translations';
import { FavoriteBooksContext } from '@/storage/FavoriteBooksContext';
import { NotificationContext } from '@/storage/NotificationContext';
import { SettingsContext } from '@/storage/SettingsContext';

type BookCardWatchListProps = {
    book: BookType;
};

export function BookCardWatchList({ book }: BookCardWatchListProps) {
    const { removeFavorite } = useContext(FavoriteBooksContext);
    const { addNotification, removeNotification, getNotificationByBookId } = useContext(NotificationContext);
    const { applicationLanguage } = useContext(SettingsContext);

    const notification: NotificationType | undefined = getNotificationByBookId(book.id);

    const getNotified = () => {
        const alertContent = i18n.t('book_notification.content', {
            bookTitle: book.title,
            bookReleaseDateRaw: formatDateStr(book.releaseDateRaw, applicationLanguage),
        });

        Alert.alert(i18n.t('book_notification.title'), alertContent, [
            {
                text: i18n.t('book_notification.cancel_button'),
                style: 'cancel',
                onPress: () => console.log('Notification declined'),
            },
            {
                text: i18n.t('book_notification.accept_button'),
                style: 'default',
                onPress: async () => {
                    scheduleBookReleaseNotification(book, addNotification);
                },
            },
        ]);
    };

    const handleToggleNotification = () => {
        if (!notification) {
            getNotified();

            return;
        }

        removeNotification(notification.id);
    };

    const handleRemoveFavorite = () => {
        Alert.alert(i18n.t('remove_favorites.title'), i18n.t('remove_favorites.content', { bookTitle: book.title }), [
            {
                text: i18n.t('remove_favorites.cancel_button'),
                style: 'cancel',
                onPress: () => {},
            },
            {
                text: i18n.t('remove_favorites.accept_button'),
                style: 'destructive',
                onPress: () => {
                    if (notification) {
                        removeNotification(notification.id);
                    }

                    removeFavorite(book.id);
                },
            },
        ]);
    };

    return (
        <View style={styles.container}>
            {/* Book Cover */}
            <Image
                source={
                    book.coverImage ? { uri: book.coverImage } : require('@/assets/images/book-cover-placeholder.jpg')
                }
                style={styles.bookCover}
                accessibilityRole="image"
                accessibilityLabel="Book cover"
            />

            {/* Book Description */}
            <View style={styles.bookDescription}>
                <ThemedText type="defaultSemiBold">{book.title}</ThemedText>
                <ThemedText type="default">{book.authors?.[0] ?? ''}</ThemedText>
                <ThemedText type="info">
                    {i18n.t('book_details.release_date', {
                        releaseDate: formatDateStr(book.releaseDateRaw, applicationLanguage),
                    })}
                </ThemedText>
            </View>

            {/* Action Icons */}
            <View style={styles.actions}>
                {!isPublished(book) && (
                    <TouchableOpacity
                        onPress={handleToggleNotification}
                        accessibilityLabel={i18n.t('book_notification.button_label')}
                    >
                        <Ionicons
                            name={notification ? 'notifications' : 'notifications-off'}
                            size={20}
                            color={notification ? Colors.gold : Colors.gray}
                        />
                    </TouchableOpacity>
                )}
                <TouchableOpacity
                    onPress={handleRemoveFavorite}
                    accessibilityLabel={i18n.t('remove_favorites.button_label')}
                >
                    <Ionicons name="trash-outline" size={20} color={Colors.red} />
                </TouchableOpacity>
            </View>
        </View>
    );
}

const styles = StyleSheet.create({
    container: {
        flexDirection: 'row',
        alignItems: 'center',
        padding: 10,
        backgroundColor: Colors.whitesmoke,
        borderRadius: 5,
        marginBottom: 5,
    },
    bookCover: {
        width: 40,
        height: 60,
    },
    bookDescription: {
        flex: 1,
        paddingLeft: 10,
    },
    actions: {
        width: 30,
        gap: 5,
        alignItems: 'center',
    },
});
