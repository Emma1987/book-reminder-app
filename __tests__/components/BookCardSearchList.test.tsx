import React from 'react';
import { Alert } from 'react-native';
import { render, fireEvent, waitFor } from '@testing-library/react-native';
import { littleSecretsBook, veryAnticipatedBook } from '@/__tests__/__fixtures__/fixtures';
import { BookCardSearchList } from '@/components/BookCardSearchList';
import { scheduleBookReleaseNotification } from '@/helpers/BookNotificationHelper';
import { formatDateStr } from '@/helpers/DateHelper';
import { FavoriteBooksContext } from '@/storage/FavoriteBooksContext';
import { NotificationContext } from '@/storage/NotificationContext';

jest.mock('@/helpers/BookNotificationHelper', () => ({
    scheduleBookReleaseNotification: jest.fn(),
}));
jest.spyOn(Alert, 'alert');

describe('BookCardSearchList', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });

    it('renders book information correctly', () => {
        const { getByText } = render(
            <FavoriteBooksContext.Provider value={{addFavorite: jest.fn(), removeFavorite: jest.fn(), isFavorite: jest.fn()}}>
                <NotificationContext.Provider value={{addNotification: jest.fn()}}>
                    <BookCardSearchList book={littleSecretsBook} onSeeDetails={jest.fn()} />
                </NotificationContext.Provider>
            </FavoriteBooksContext.Provider>
        );

        expect(getByText('Little Secrets')).toBeTruthy();
        expect(getByText('Jennifer Hillier')).toBeTruthy();
        expect(getByText('Release date: 7 May 2020')).toBeTruthy();
    });

    it('renders only first author when several authors', () => {
        const mockBookWithSeveralAuthors = {
            ...littleSecretsBook,
            authors: ['Stephen King', 'Jennifer Hillier'],
        };

        const { getByText } = render(
            <FavoriteBooksContext.Provider value={{addFavorite: jest.fn(), removeFavorite: jest.fn(), isFavorite: jest.fn()}}>
                <NotificationContext.Provider value={{addNotification: jest.fn()}}>
                    <BookCardSearchList book={mockBookWithSeveralAuthors} onSeeDetails={jest.fn()} />
                </NotificationContext.Provider>
            </FavoriteBooksContext.Provider>
        );

        expect(getByText('Little Secrets')).toBeTruthy();
        expect(getByText('Stephen King')).toBeTruthy();
        expect(getByText('Release date: 7 May 2020')).toBeTruthy();
    });

    it('renders placeholder image when coverImage is null', () => {
        const mockBookWithoutCover = {
            ...littleSecretsBook,
            coverImage: null,
        };
        
        const { getByLabelText } = render(
            <FavoriteBooksContext.Provider value={{addFavorite: jest.fn(), removeFavorite: jest.fn(), isFavorite: jest.fn()}}>
                <NotificationContext.Provider value={{addNotification: jest.fn()}}>
                    <BookCardSearchList book={mockBookWithoutCover} onSeeDetails={jest.fn()} />
                </NotificationContext.Provider>
            </FavoriteBooksContext.Provider>
        );

        const image = getByLabelText('Book cover');
        expect(image.props.source).toEqual(require('@/assets/images/book-cover-placeholder.jpg'));
    });

    it('renders book cover when coverImage is provided', () => {
        const { getByLabelText } = render(
            <FavoriteBooksContext.Provider value={{addFavorite: jest.fn(), removeFavorite: jest.fn(), isFavorite: jest.fn()}}>
                <NotificationContext.Provider value={{addNotification: jest.fn()}}>
                    <BookCardSearchList book={littleSecretsBook} onSeeDetails={jest.fn()} />
                </NotificationContext.Provider>
            </FavoriteBooksContext.Provider>
        );
    
        const image = getByLabelText('Book cover');
        expect(image.props.source).toEqual({ uri: 'https://book.com/little-secrets-cover.jpg' });
    });

    it('adds book to favorite when favorite button is clicked and book is not favorite', () => {
        const mockAddFavorite = jest.fn();
        const mockIsFavorite = jest.fn();
        mockIsFavorite.mockReturnValue(false);

        const { getByLabelText } = render(
            <FavoriteBooksContext.Provider value={{addFavorite: mockAddFavorite, removeFavorite: jest.fn(), isFavorite: mockIsFavorite}}>
                <NotificationContext.Provider value={{addNotification: jest.fn()}}>
                    <BookCardSearchList book={littleSecretsBook} onSeeDetails={jest.fn()} />
                </NotificationContext.Provider>
            </FavoriteBooksContext.Provider>
        );

        const favoriteButton = getByLabelText('Add to favorites');
        fireEvent.press(favoriteButton);

        expect(mockAddFavorite).toHaveBeenCalledWith(littleSecretsBook);
    });

    it('removes book from favorites when favorite button is clicked and book is already favorite', () => {
        const mockRemoveFavorite = jest.fn();
        const mockIsFavorite = jest.fn();
        mockIsFavorite.mockReturnValue(true);

        const { getByLabelText } = render(
            <FavoriteBooksContext.Provider value={{addFavorite: jest.fn(), removeFavorite: mockRemoveFavorite, isFavorite: mockIsFavorite}}>
                <NotificationContext.Provider value={{addNotification: jest.fn()}}>
                    <BookCardSearchList book={littleSecretsBook} onSeeDetails={jest.fn()} />
                </NotificationContext.Provider>
            </FavoriteBooksContext.Provider>
        );

        const favoriteButton = getByLabelText('Add to favorites');
        fireEvent.press(favoriteButton);

        expect(mockRemoveFavorite).toHaveBeenCalledWith(littleSecretsBook.id);
    });

    it('calls onSeeDetails when details button is clicked', () => {
        const mockOnSeeDetails = jest.fn();

        const { getByLabelText } = render(
            <FavoriteBooksContext.Provider value={{addFavorite: jest.fn(), removeFavorite: jest.fn(), isFavorite: jest.fn()}}>
                <NotificationContext.Provider value={{addNotification: jest.fn()}}>
                    <BookCardSearchList book={littleSecretsBook} onSeeDetails={mockOnSeeDetails} />
                </NotificationContext.Provider>
            </FavoriteBooksContext.Provider>
        );

        const detailsButton = getByLabelText('View book details');
        fireEvent.press(detailsButton);

        expect(mockOnSeeDetails).toHaveBeenCalledWith(littleSecretsBook);
    });

    it('has correct accessibility labels for the buttons', () => {
        const { getByLabelText } = render(
            <FavoriteBooksContext.Provider value={{addFavorite: jest.fn(), removeFavorite: jest.fn(), isFavorite: jest.fn()}}>
                <NotificationContext.Provider value={{addNotification: jest.fn()}}>
                    <BookCardSearchList book={littleSecretsBook} onSeeDetails={jest.fn()} />
                </NotificationContext.Provider>
            </FavoriteBooksContext.Provider>
        );

        const favoriteButton = getByLabelText('Add to favorites');
        expect(favoriteButton).toBeTruthy();

        const detailsButton = getByLabelText('View book details');
        expect(detailsButton).toBeTruthy();
    });

    it('shows alert when favoriting an unpublished book', () => {
        const { getByLabelText } = render(
            <FavoriteBooksContext.Provider value={{addFavorite: jest.fn(), removeFavorite: jest.fn(), isFavorite: jest.fn()}}>
                <NotificationContext.Provider value={{addNotification: jest.fn()}}>
                    <BookCardSearchList book={veryAnticipatedBook} onSeeDetails={jest.fn()} />
                </NotificationContext.Provider>
            </FavoriteBooksContext.Provider>
        );

        fireEvent.press(getByLabelText('Add to favorites'));

        expect(Alert.alert).toHaveBeenCalledWith(
            'Stay Updated!',
            expect.stringContaining(`The book "Very anticipated book!" is set to be released on ${formatDateStr(veryAnticipatedBook.releaseDateRaw, 'monthFirst')}`),
            expect.any(Array) // Ensures alert buttons are present
        );
    });

    it('schedules notification when clicking "Yes, notify me!"', async () => {
        const mockAddNotification = jest.fn();

        const { getByLabelText } = render(
            <FavoriteBooksContext.Provider value={{addFavorite: jest.fn(), removeFavorite: jest.fn(), isFavorite: jest.fn()}}>
                <NotificationContext.Provider value={{addNotification: mockAddNotification}}>
                    <BookCardSearchList book={veryAnticipatedBook} onSeeDetails={jest.fn()} />
                </NotificationContext.Provider>
            </FavoriteBooksContext.Provider>
        );

        fireEvent.press(getByLabelText('Add to favorites'));

        const alertButtons = Alert.alert.mock.calls[0][2];

        // Simulate clicking "Yes, notify me!"
        await waitFor(() => alertButtons[1].onPress());

        fireEvent.press(getByLabelText('Add to favorites'));

        expect(scheduleBookReleaseNotification).toHaveBeenCalledTimes(1);
    });

    it('does not schedule notification when clicking "No, thanks"', () => {
        const { getByLabelText } = render(
            <FavoriteBooksContext.Provider value={{addFavorite: jest.fn(), removeFavorite: jest.fn(), isFavorite: jest.fn()}}>
                <NotificationContext.Provider value={{addNotification: jest.fn()}}>
                    <BookCardSearchList book={veryAnticipatedBook} onSeeDetails={jest.fn()} />
                </NotificationContext.Provider>
            </FavoriteBooksContext.Provider>
        );

        fireEvent.press(getByLabelText('Add to favorites'));

        const alertButtons = Alert.alert.mock.calls[0][2];

        // Simulate clicking "No, thanks"
        alertButtons[0].onPress();

        expect(scheduleBookReleaseNotification).not.toHaveBeenCalled();
    });

    it('matches the snapshot', () => {
        const tree = render(
            <FavoriteBooksContext.Provider value={{addFavorite: jest.fn(), removeFavorite: jest.fn(), isFavorite: jest.fn()}}>
                <NotificationContext.Provider value={{addNotification: jest.fn()}}>
                    <BookCardSearchList book={littleSecretsBook} onSeeDetails={jest.fn()} />
                </NotificationContext.Provider>
            </FavoriteBooksContext.Provider>
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});
