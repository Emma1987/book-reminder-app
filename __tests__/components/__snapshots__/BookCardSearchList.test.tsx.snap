// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookCardSearchList matches the snapshot 1`] = `
<View
  style={
    {
      "borderBottomWidth": 1,
      "borderColor": "white",
      "padding": 10,
    }
  }
>
  <View
    accessibilityHint="View more details about "Little Secrets""
    accessibilityLabel="View more details about "Little Secrets""
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "alignItems": "center",
        "flexDirection": "row",
        "opacity": 1,
      }
    }
  >
    <Image
      accessibilityLabel="Book cover"
      accessibilityRole="image"
      source={
        {
          "uri": "https://book.com/little-secrets-cover.jpg",
        }
      }
      style={
        {
          "height": 90,
          "width": 60,
        }
      }
    />
    <View
      style={
        {
          "flex": 1,
          "padding": 10,
        }
      }
    >
      <Text
        style={
          [
            undefined,
            undefined,
            undefined,
            {
              "fontSize": 20,
              "fontWeight": "bold",
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
          ]
        }
      >
        Little Secrets
      </Text>
      <Text
        style={
          [
            undefined,
            undefined,
            {
              "fontSize": 16,
              "fontWeight": "600",
              "lineHeight": 24,
            },
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
            undefined,
          ]
        }
      >
        Jennifer Hillier
      </Text>
      <Text>
        Release date: May 7, 2020
      </Text>
    </View>
    <View
      style={
        {
          "width": 40,
        }
      }
    >
      <View
        accessibilityHint="Add this book to your favorites"
        accessibilityLabel="Add this book to your favorites"
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "opacity": 1,
          }
        }
      >
        <Text>
          heart-outline
        </Text>
      </View>
    </View>
  </View>
</View>
`;
